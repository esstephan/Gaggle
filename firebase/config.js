import firebase from 'firebase';

let config = {
  apiKey: "",
  authDomain: "",
  databaseURL: "",
  storageBucket: "",
  messagingSenderId: ""
};

export default firebaseRef = firebase.initializeApp(config);