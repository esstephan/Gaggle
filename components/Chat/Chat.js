import React, { Component } from 'react';
import { StyleSheet, View, TextInput } from 'react-native';
import { Container, Footer, Content, Icon, Button, ListItem, List, Text } from 'native-base';
import { firebaseDB } from '../../firebase/config.js';
import moment from 'moment';

export default class Chat extends Component {
  constructor(props, context) {
    super(props, context);
    this._handleChangePage = this._handleChangePage.bind(this);
    this.database = firebaseDB;
    this.state = {
      username: this.props.username,
      input: '',
      group: this.props.groupName ? this.props.groupName : 'Default',
      messages: []
    }

    this.messagesRef = this.database.ref('messages/' + this.state.group);
    this.sendMessage = this.sendMessage.bind(this);
  }

  componentDidMount() {
    this.messagesListener();
  }

  _handleChangePage() {
    this.props.navigator.pop();
  }

  messagesListener() {
    this.messagesRef.on('value', (snapshot) => {
      // Handle no messages created yet...
      if (snapshot.val() === null) {
        this.setState({messages: []});
      } else {
        this.setState({messages: snapshot.val()});
      }
    });
  }

  sendMessage() {
    // Write a message into database
    // Transaction will allow firebase to queue the requests so messages aren't written at the same time
    this.messagesRef.transaction((messages) => {
      if (!messages) {
        messages = [];
      }
      messages.push({
        name: this.state.username,
        message: this.state.input,
        timestamp: new Date().getTime()
      });

      // Clear TextInput
      this.setState({input: ''});

      return messages;
    });
  }

  render() {
    return (
      <Container>
        <Content>
          <List dataArray={this.state.messages} renderRow={(obj) =>
            <ListItem>
              <Text>{obj.name} ({moment(obj.timestamp).fromNow()}): </Text>
              <Text>{obj.message}</Text>
            </ListItem>
          } />
        </Content>
        <Footer style={{position: 'absolute', top: 500}}>
          <TextInput style={styles.textInput}
            value={this.state.input}
            onChangeText={(t) => this.setState({input: t})}
          />
          <View style={{position: 'relative', top: 12}}>
            <Button small onPress={this.sendMessage}>
              <Text style={{color: 'white'}}>Send</Text>
            </Button>
          </View>
          <View style={{paddingRight: 10}}></View>
        </Footer>
      </Container>
    );
  }
}

const styles = StyleSheet.create({
  textInput:{
    flex:1,
    borderColor: 'grey',
    borderWidth: 1,
    paddingLeft: 10,
    margin: 10
  }
});
