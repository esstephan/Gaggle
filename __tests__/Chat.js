import 'react-native';
import React from 'react';
import Chat from '../components/Chat/Chat.js';
import { firebaseRef } from '../firebase/config.js';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

xit('renders correctly', () => {
  const tree = renderer.create(
    <Chat />
  );
  expect(tree).toMatchSnapshot();
});
