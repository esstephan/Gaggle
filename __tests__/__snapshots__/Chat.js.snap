exports[`test renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "height": 500,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      virtual={undefined} />
    <View>
      <RCTScrollView>
        <View />
      </RCTScrollView>
    </View>
  </View>
  <View
    style={
      Object {
        "borderTopColor": "lightgrey",
        "borderTopWidth": 1,
        "flex": 1,
        "flexDirection": "row",
      }
    }>
    <View
      style={
        Object {
          "flex": 4,
          "height": 50,
        }
      }>
      <TextInput
        onChangeText={[Function]}
        style={
          Object {
            "borderColor": "grey",
            "borderWidth": 1,
            "flex": 1,
            "margin": 10,
            "paddingLeft": 10,
          }
        }
        value="" />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 10,
        }
      }>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "flex-start",
            "backgroundColor": "#007aff",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "elevation": 2,
            "flexDirection": "row",
            "height": 30,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 6,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}>
        <View
          style={
            Object {
              "alignSelf": "center",
              "zIndex": 999,
            }
          } />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
            }
          }
          virtual={undefined}>
          Send
        </Text>
      </View>
    </View>
  </View>
</View>
`;
